pkg/apis/workflow/v1alpha1/workflow_types.go

```diff
@@ -869,7 +872,7 @@ func (s *S3Artifact) String() string {
 }
 
 func (s *S3Artifact) HasLocation() bool {
-	return s != nil && s.Bucket != ""
+	return s != nil && (s.Bucket != "" || s.BucketSecret.Key != "")
```

Comment:
is it possible here that `s.BucketSecret` could be `nil`?

Reply:
I don't think so, swagger seems to be always initializing BucketSecret, but with empty variables

Reply:
I think that `bucketSecret` should be nillable.

Reply:
if you make it nillable, checking accessing it's keys will always be risky of raising hard to debug exceptions. Way it's done now is that Key will default to "" if no bucketSecret is provided in config, so it's easier to handle. Also I remember having issues with swagger json def and nillable values. I couldn't define #/definitions/io.k8s.api.core.v1.SecretKeySelector or nil in swagger.json

Reply:
@jessesuen suggested that it should not be nillable if there is a zero value, e.g. string=="" of bool=false, would I be right in thinking that's the case for bucketSecret?

Reply:
I don't understand the question. If you ask if there is default value that it will take if person doesn't specify `bucketSecret` in config at all, then answer is yes. If you don't have `bucketSecret` in config, Key will be "" and everything works as intended.

Reply:
Item 1.

I think my query comes from the swagger.json. If you can have either bucket or bucketKey, then they should not be required in the swagger.json. So i think they should have `omitempty` on them.

Item 2.

I wanted to check that it would be possible to omit them. I think you have confirm this.